src/ucs/sys/sock.c

```diff
@@ -366,14 +370,33 @@ ucs_status_t ucs_socket_server_init(const struct sockaddr *saddr, socklen_t sock
         goto err_close_socket;
     }
 
-    status = ucs_sockaddr_get_port(saddr, &port);
-    if (status != UCS_OK) {
-        goto err_close_socket;
+    if ((ports_range_start != 0) && (ports_range_start == ports_range_end)) {
+        /* non zero single value port */
+        retry = 0;
+    } else {
+        /* port value is zero or a range of ports is given */
+        retry = 1;
     }
 
+    rand_seed = ucs_generate_uuid((uintptr_t)saddr);
+
     do {
+        if ((ports_range_start != 0) || (ports_range_end != 0)) {
+            port = (uint16_t)((rand_r(&rand_seed) %
+                              (ports_range_end -
+                               ports_range_start + 1)) +
+                              ports_range_start);
+        } else {
+            port = 0;   /* use a random port */
+        }
+
+        status = ucs_sockaddr_set_port((struct sockaddr *)saddr, port);
+        if (status != UCS_OK) {
+            goto err_close_socket;
+        }
+
         ret = bind(fd, saddr, socklen);
-    } while (!port && (ret < 0) && (errno == EADDRINUSE));
+    } while (retry && (ret < 0) && (errno == EADDRINUSE));
```

Comment:
instead of retry can we check (port == 0) ?

Reply:
we need to retry both when port is 0 or if it's not zero in a range.

Reply:
when we don't need to retry?

Reply:
according to the code these conditions are equivalent

Reply:
we don't need to retry when there is a single non zero value - like it might come from tcp_sockcm listener.

Reply:
ok